yapapi/runner/events.py

```diff
@@ -0,0 +1,117 @@
+"""Code for representing and logging events in Golem computation."""
+import asyncio
+from dataclasses import dataclass, field
+from enum import Enum, auto
+from typing import Any, ClassVar, Mapping, Optional, TypeVar, Union
+
+from typing_extensions import Protocol
+
+
+ResourceId = Union[int, str]
+EventType = str
+
+
+class ResourceType(Enum):
+    """Resource types for events."""
+
+    SUBSCRIPTION = auto()
+    PROPOSAL = auto()
+    AGREEMENT = auto()
+    ACTIVITY = auto()
+    WORKER = auto()
+    TASK = auto()
+
+
+@dataclass(frozen=True)
+class _Event:
+    """Represents an event in a Golem computation."""
+
+    resource_type: ResourceType
+    event_type: EventType
+    resource_id: Optional[ResourceId] = None
+    info: Mapping[str, Any] = field(default_factory=dict)
+
+    event_type_to_readable: ClassVar[Mapping[str, str]] = {
+        "recv": "received",
+        "fail-subscribe": "failed to subscribe to market events",
+        "fail-collect": "failed to collect proposal events",
+        "fail-create": "could not be created on provider",
+        "fail-run": "cannot run commands on provider",
+        "respond": "answered",
+        "get-work": "getting work",
+        "get-results": "getting batch results",
+        "create": "created",
+        "batch-done": "batch done",
+        "confirm": "confirmed and approved",
+        "fail": "failed",
+        "reject": "rejected",
+        "accept": "accepted",
+        "payment_prep": "related payment prepared",
+        "payment_queued": "related payment queued",
+        "payment_accept": "related payment accepted",
+        "downloading": "downloading file",
+        "downloaded": "downloaded file",
+        "new-batch": "(ExeUnit) script with commands sent",
+        "to-pay": "ids to pay for:",
+    }
+
+    def __str__(self) -> str:
+        """Return a human-readable representation of the event."""
+
+        readable_type = self.event_type_to_readable.get(
+            self.event_type, self.event_type
+        )
+        return (
+            f"[{self.resource_type.name.title()}] {readable_type}, "
+            f"id = {self.resource_id}, info = {self.info}"
+        )
+
+
+R = TypeVar("R", contravariant=True, bound=ResourceType)
+
+
+class EventEmitter(Protocol[R]):
+    """A protocol for callback functions that emit events."""
+
+    def __call__(
+        self,
+        resource_type: R,
+        event_type: EventType,
+        resource_id: Optional[ResourceId] = None,
+        **kwargs: Any
+    ) -> None:
+        """Emit an event."""
+
+
+class AsyncEventLogger:
```

Comment:
@azawlocki wondering aloud -> shouldn't this class be a singleton? e.g. in case we would like to somehow attach independent listeners to some events? maybe then it would be better not to have to dig to get the specific event logger but rather have a go-to instance? ...

Reply:
@shadeofblue Hmm, I don't think so. In the original code, the task that emits events was created in `Runner.map()`. A reference to it was also passed to `Task.start()`. This suggests that many instances should be supported and I wouldn't like to change this without clear reason.

But pehaps the name `Logger` suggests a singleton? If you think so then I may consider changing it.

Reply:
okay, gotcha

and no - I don't think the name is the issue here ;)